cmd/snap/cmd_quota.go

```diff
@@ -132,3 +142,74 @@ type cmdRemoveQuota struct {
 func (x *cmdRemoveQuota) Execute(args []string) (err error) {
 	return x.client.RemoveQuotaGroup(x.Positional.GroupName)
 }
+
+type cmdQuotas struct {
+	clientMixin
+}
+
+func (x *cmdQuotas) Execute(args []string) (err error) {
+	res, err := x.client.Quotas()
+	if err != nil {
+		return err
+	}
+	if len(res) == 0 {
+		fmt.Fprintln(Stdout, i18n.G("No quota groups defined."))
+		return nil
+	}
+
+	w := tabWriter()
+	fmt.Fprintf(w, "Quota\tParent\tMax-Memory\n")
+	processQuotaGroupsTree(res, func(q *client.QuotaGroupResult) {
+		fmt.Fprintf(w, "%s\t%s\t%s\n", q.GroupName, q.Parent, fmtSize(int64(q.MaxMemory)))
+	})
+	w.Flush()
+	return nil
+}
+
+type quotaGroup struct {
+	res       *client.QuotaGroupResult
+	subGroups []*quotaGroup
+}
+
+type byQuotaName []*quotaGroup
+
+func (q byQuotaName) Len() int           { return len(q) }
+func (q byQuotaName) Swap(i, j int)      { q[i], q[j] = q[j], q[i] }
+func (q byQuotaName) Less(i, j int) bool { return q[i].res.GroupName < q[j].res.GroupName }
+
+// processQuotaGroupsTree recreates the hierarchy of quotas and then visits it
+// recursively following the hierarchy first, then naming order.
+func processQuotaGroupsTree(quotas []*client.QuotaGroupResult, handleGroup func(q *client.QuotaGroupResult)) {
+	var roots []*quotaGroup
+	groupLookup := make(map[string]*quotaGroup, len(quotas))
+
+	for _, q := range quotas {
+		grp := &quotaGroup{res: q}
+		groupLookup[q.GroupName] = grp
+
+		if q.Parent == "" {
+			roots = append(roots, grp)
+		}
+	}
+
+	sort.Sort(byQuotaName(roots))
+
+	// populate sub-groups
+	for _, g := range groupLookup {
+		sort.Strings(g.res.Subgroups)
+		for _, subgrpName := range g.res.Subgroups {
+			g.subGroups = append(g.subGroups, groupLookup[subgrpName])
```

Comment:
Should we try to reasonably deal with inconsistency in the input if the subgroup was not present in quotas list and signal an error (followed by dump of data)?

Reply:
Not sure, that would be a bug / internal error in snapd already, we ensure consistency when groups are created.

Reply:
I'll address this concern in a followup.